src/staging-heap.cpp

```diff
@@ -5,25 +5,28 @@
 
 namespace rhi {
 
-void StagingHeap::initialize(Device* device)
+void StagingHeap::initialize(Device* device, Size pageSize)
 {
+    std::lock_guard<std::mutex> lock(m_mutex);
     m_device = device;
+    m_pageSize = pageSize;
 }
 
 void StagingHeap::releaseAllFreePages()
 {
-    std::vector<int> pagesToRemove;
+    std::vector<Page*> pagesToRemove;
     for (auto& page : m_pages)
     {
         if (page.second->getUsed() == 0)
-            pagesToRemove.push_back(page.first);
+            pagesToRemove.push_back(page.second.get());
     }
-    for (auto page_id : pagesToRemove)
-        m_pages.erase(page_id);
+    for (auto page : pagesToRemove)
+        freePage(page);
 }
 
 void StagingHeap::release()
 {
+    std::lock_guard<std::mutex> lock(m_mutex);
```

Comment:
again, not sure this needs a lock .. is `StagingHeap` supposed to be initialized/released from multiple threads?